pkg/registry/core/service/strategy.go

```diff
@@ -187,14 +183,19 @@ func dropServiceDisabledFields(newSvc *api.Service, oldSvc *api.Service) {
 }
 
 // returns true if svc.Spec.ServiceIPFamily field is in use
-func serviceIPFamilyInUse(svc *api.Service) bool {
+func (strategy svcStrategy) serviceDualStackFieldsInUse(svc *api.Service) bool {
 	if svc == nil {
 		return false
 	}
-	if svc.Spec.IPFamily != nil {
-		return true
-	}
-	return false
+
+	ipFamilyPolicyInUse := svc.Spec.IPFamilyPolicy != nil && *(svc.Spec.IPFamilyPolicy) != api.SingleStack
+	// we have to check if user is using a family not configured on cluster in single stack gate off mode
+	// in order to return correct errors.
+	ipFamiliesInUse := len(svc.Spec.IPFamilies) == 2 ||
```

Comment:
We are not setting or using them. If the user is using them (let us say gate was on at some point) and data had something that can only happen when the gate is on, then turned off the gate.. if we don't check for that then we will drop these fields (including trimming ClusterIPs to ClusterIPs[0] which basically modifies data the user previously entered.

Reply:
The allocator is checked for gates as it loads (it is based on cluster configuration and so on). It will not allocate for an alternative/secondary family that is not configured on cluster. This code must work in a consistent way irrespective of gate status (because it can be used to clear field if the gate was on and then turned off.